PEP 617: Clean-up of LL(1) section (#1363)
python/peps: pep-0617.rst

Syntax Tree (CST) sometimes known directly as a "parse tree". In this process, the
*first sets* are used indirectly when generating the DFAs.

LL(1) parsers and grammars are usually efficient and simple to implement
and generate. However, it is not possible, under the LL(1) restriction,
to express certain common constructs in a way natural to the language
designer and the reader. This includes some in the Python language.

As LL(1) parsers can only look one token ahead to distinguish
possibilities, some rules in the grammar may be ambiguous. For instance the rule::

    rule: A | B

is ambiguous if the *first sets* of both ``A`` and ``B`` have some elements in
common. When the parser sees a token in the input
program that both *A* and *B* can start with, it is impossible for it to deduce
which option to expand, as no further token of the program can be examined to
disambiguate.
The rule may be transformed to equivalent LL(1) rules, but then it may
be harder for a human reader to grasp its meaning.
Examples later in this document show that the current LL(1)-based
grammar suffers a lot from this scenario.

> Another broad class of rule precluded by LL(1) is the left-recursive rule.

Reviewer: ```suggestion
Another broad class of rules precluded by LL(1) is left-recursive rules.
```

Author: I disagree with the first plural ("class of rule" sounds right to me) but I realise [I'm in the minority](https://english.stackexchange.com/questions/5539/types-of-things-vs-types-of-thing). The rest is equivalent or better.

> A rule is left-recursive if there exists a nonterminal that can derive to a

Reviewer: ```suggestion
A rule is left-recursive if it can derive to a
```

Author: Quite right. Hang-over from "a grammar is LR if ..."
